app/controllers/users_controller.rb

```diff
@@ -46,7 +46,12 @@ def edit
   end
 
   def update
-    if current_user.update(updated_params)
+    current_user.batch_id = -1 if params.dig(:user, :batch_number) # Set impossible value to trigger validation
+
+    referrer_code = params.dig(:user, :referrer_code)
+    current_user.referrer_id = User.find_by_referral_code(referrer_code)&.id || -1 if referrer_code
+
+    if current_user.update(user_params)
```

Comment:
```suggestion
    if current_user.update(user_params)
```

On a pas besoin de ces histoires de `-1` avec les validations dans `User` il me semble. Quelque chose m'échappe ?

Reply:
Si un user ouvre le DevTool et enleve manuellement le disable du bouton pour le batch, rentre la valeur qu'il veut et submit, on veut quand meme pouvoir lui mettre un message d'erreur.
D'ou
```ruby
current_user.batch_id = -1 if params.dig(:user, :batch_number)
```

Et si un user veut changer son referrer pendant le setup il a besoin de pouvoir rentré un referral_code, d'ou
```ruby
    referrer_code = params.dig(:user, :referrer_code)
    current_user.referrer_id = User.find_by_referral_code(referrer_code)&.id if referrer_code
```
Sauf que si il rentre un code erroné on veut quand meme pouvoir validé, donc on ajoute `|| -1` si le code est erroné

Reply:
> Si un user ouvre le DevTool et enleve manuellement le disable du bouton pour le batch, rentre la valeur qu'il veut et submit, on veut quand meme pouvoir lui mettre un message d'erreur.

Il se ferait tej par [cette validation](https://github.com/pil0u/lewagon-aoc/pull/318/files#diff-9802ca3c9c4cf89904fd44bc114e35ebdf2c5dd3d5b645491e2b253e1afef29bR31) normalement.

> Sauf que si il rentre un code erroné on veut quand meme pouvoir validé, donc on ajoute || -1 si le code est erroné

Ça en effet, mais on pourrait mieux le gérer via un `attribute :referrer_code` qui n'existe pas en DB mais sert à ce genre de validations.

Reply:
> Il se ferait tej par [cette validation](https://github.com/pil0u/lewagon-aoc/pull/318/files#diff-9802ca3c9c4cf89904fd44bc114e35ebdf2c5dd3d5b645491e2b253e1afef29bR31) normalement.
Non parce que le param est pas permitted.
On trouvais pas ca logiquer de permit un param qu'on va jamais autoriser alors plutot que de l'ajouter dans le `require.permit` on s'est dit que ca faisait plus de sens de le mettre dans l'update .
Et du coup quitte a l'assigner a qqch Pilou preferait mettre `-1` pour que ca soit bien clair que c'etait impossible

Reply:
> on pourrait mieux le gérer via un attribute :referrer_code

pas la ref

Reply:
Pas de soucis. Je verrai bien plus tard si on peut clean ça.